Profiling on Processors level

azat opened the pull request:
Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Profiling on Processors level (under `log_processors_profiles` setting, ClickHouse will write time that processor spent during execution/waiting for data to `system.processors_profile_log` table)

Fixes: #15261
Cc: @alexey-milovidov 
Cc: @KochetovNicolai 

*I've looked for something similar and found this issue.*

azat committed:
Add system.processors_profile_log

This is the system table that will contain Processors level profiling.

v2: one entry per Processor, not 3 (PortFull/NeedData/work())
v3: us over ms
v4: Enable processors_profile_log table by default

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Implement log_processors_profiles (write to processors_profile_log, OFF by default)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Measure time that Process spent in work/NeedData/PortFull

Note, that right now it is done not in IProcessor, but in
ExecutingGraph/ExecutionThreadContext, to avoid lots of changes in the
IProcessor interface, to make review easier, but I'm not against of
change the IProcessor interface to incapsulate it there.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: cover processors_profile_log

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented on src/Processors/IProcessor.h:
```diff
@@ -299,14 +300,33 @@ class IProcessor
+    /// For
+    /// - need_data_elapsed_us
+    /// - port_full_elapsed_us
+    friend class ExecutingGraph;
```
Note, that right now, I decided to make measurements in the callers instead of in IProcessor itself, this is to avoid lots of changes in the interface, to make review easier.
But I'm not against of change the IProcess interface to encapsulate the logic in it.

KochetovNicolai replied:
Yes, I would definitely not touch the processors's interface.

KochetovNicolai replied:
I would actually do not add anything new to IProcessor

azat replied:
Extracting this information may be too complex, I think putting it into processors is OK, plus it will allow to see these metrics in debugger too.
Objections?

KochetovNicolai commented on docs/en/operations/system-tables/processors_profile_log.md:
```diff
@@ -0,0 +1,73 @@
+-   `event_date` ([Date](../../sql-reference/data-types/date.md)) — The date when the event happened.
+-   `event_time` ([DateTime64](../../sql-reference/data-types/datetime64.md)) — The date and time when the event happened.
+-   `query_id` ([String](../../sql-reference/data-types/string.md)) — ID of the query
+-   `name` ([String](../../sql-reference/data-types/string.md)) — Name of the processor.
```
Well, many processors may have similar names. I suppose, we add a separate event for each processor, still it's hard to understand which one is responsible for what.

Don't know what should be a better solution. Maybe, enumerate processors for a query, and also add info about the graph structure. Like, parents (or children) array.

azat replied:
Yeah, name is too short to identify.
Although graph structure sounds complicated for these table.
Maybe description will be enough?
I will try description first and if it will not be enough, then I will add something more.

KochetovNicolai replied:
For each processor we can just store an array of parents. I think it won't be much heavier.
It would be possible to build a graph with a query from this table.

azat replied:
Done.

KochetovNicolai commented on docs/en/operations/system-tables/processors_profile_log.md:
```diff
@@ -0,0 +1,73 @@
+-   `query_id` ([String](../../sql-reference/data-types/string.md)) — ID of the query
+-   `name` ([String](../../sql-reference/data-types/string.md)) — Name of the processor.
+-   `elapsed_us` ([UInt64](../../sql-reference/data-types/int-uint.md)) — Number of microseconds this processor was executed.
+-   `need_data_elapsed_us` ([UInt64](../../sql-reference/data-types/int-uint.md)) — Number of microseconds this processor was waiting for data (from other processor).
+-   `port_full_elapsed_us` ([UInt64](../../sql-reference/data-types/int-uint.md)) — Number of microseconds this processor was waiting because output port was full.
```
Bad names. Would better to sum those together and call like `io_wait_ns`. Or maybe separate `input_wait_ns` and `output_wait_ns`.

azat replied:
Indeed, input_wait_us/output_wait_us sounds way more better, thanks!

KochetovNicolai commented on src/Processors/Executors/ExecutingGraph.cpp:
```diff
@@ -263,7 +263,30 @@ bool ExecutingGraph::updateNode(uint64_t pid, Queue & queue, Queue & async_queue
+                    /// NeedData
+                    if (last_status != IProcessor::Status::NeedData && status == IProcessor::Status::NeedData)
+                    {
+                        processor.need_data_watch.restart();
```
Note that similar stopwatch is under define only for DEBUG. I've seen some (a few percent) slow down for some queries because of gettime. We should only measure time if setting is enabled.

azat replied:
>Note that similar stopwatch is under define only for DEBUG. I've seen some (a few percent) slow down for some queries because of gettime. 

Interesting, but perf tests does not show this, I guess this is was some synthetic query.
By the way, maybe it was before vdso had been fixed?

>We should only measure time if setting is enabled.

Ok.

azat replied:
By the way, I guess one day it will be enabled by default...

azat replied:
I did this, and I don't like the complexity that it introduces.

After I've looked at the overhead and I don't see any difference (for `select count() from numbers(1e9)` query):

- patched

<details>

```sh
Queries executed: 100.

localhost:9000, queries 100, QPS: 3.958, RPS: 3958441381.337, MiB/s: 30200.511, result RPS: 3.958, result MiB/s: 0.000.

0.000%          0.238 sec.
10.000%         0.241 sec.
20.000%         0.244 sec.
30.000%         0.246 sec.
40.000%         0.248 sec.
50.000%         0.249 sec.
60.000%         0.252 sec.
70.000%         0.255 sec.
80.000%         0.263 sec.
90.000%         0.269 sec.
95.000%         0.274 sec.
99.000%         0.280 sec.
99.900%         0.285 sec.
99.990%         0.285 sec.
```

</details>

- upstream

<details>

```sh
Queries executed: 100.

localhost:9000, queries 100, QPS: 3.826, RPS: 3825878323.592, MiB/s: 29189.135, result RPS: 3.826, result MiB/s: 0.000.

0.000%          0.244 sec.
10.000%         0.253 sec.
20.000%         0.257 sec.
30.000%         0.258 sec.
40.000%         0.259 sec.
50.000%         0.261 sec.
60.000%         0.262 sec.
70.000%         0.264 sec.
80.000%         0.266 sec.
90.000%         0.270 sec.
95.000%         0.274 sec.
99.000%         0.286 sec.
99.900%         0.287 sec.
99.990%         0.287 sec.
```

</details>

So my suggestion is to add this code only if there will be a proof that it makes the difference.

azat replied:
Added in https://github.com/ClickHouse/ClickHouse/pull/34355/commits/5475155f88b94d0be163434be182009e1206b343, but I'm going to remove it if there will be no objections, since I don't like the complexity that it adds.

azat commented:
>I still don't know if it is a really good idea to add system table for processors profiling. Maybe it would be better to use OpenTelemetry as a more general approach (e.g. show the full execution timeline for each processor, not just sum time).
>
>Maybe it is still useful in case if we need more detailed profiling, but can't setup OpenTelemetry

I think providing something, simple enough, will allow to make debugging easier, since not a lot of users will configure OpenTelemetry I guess (plus it requires external service to look into it).

Plus separate table will allow to do aggregations for particular types of queries to see where it spends more time (reading, sorting, grouping, ...)

But providing OpenTelemetry support for processors_profile sounds good.

alexey-milovidov commented:
+1 having simple system table is better than to force users to rely on OpenTelemetry.
For OpenTelemetry even its name sounds like something complicated and difficult to use.

UnamedRus commented:
I think, it make sense to include metrics like rows before and rows after for each processor step.

azat commented:
>I think, it make sense to include metrics like rows before and rows after for each processor step.

Great idea, but right now interface does not allow to do this.

azat committed:
Rename need_data_elapsed_us/port_full_elapsed_us to input_wait_us/output_wait_us

$ gg -e need_data_ -e port_full_  | cut -d: -f1 | sort -u | xargs sed -i -e s/port_full_/output_wait_/g -e s/need_data_/input_wait_/g -e s/getPortFull/getOutputWait/g -e s/getNeedData/getInputWait/g

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Mesure processors profiles only if it was enabled

Since it may use little extra CPU.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add id/parent_ids into system.processors_profile_log

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
@KochetovNicolai can you also please take a look at this?

azat commented:
@KochetovNicolai friendly ping

KochetovNicolai commented on src/Interpreters/executeQuery.cpp:
```diff
@@ -865,6 +867,44 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
+                        processor_elem.event_time_microseconds = time_in_microseconds(finish_time);
+                        processor_elem.query_id = elem.client_info.current_query_id;
+
+                        auto get_proc_id = [](const IProcessor & proc) -> UInt64
+                        {
+                            return reinterpret_cast<std::uintptr_t>(&proc);
+                        };
```
Maybe it would be good to re-enumerate processors, so that ids were in [0 ... N - 1]
I thought about it implementing `explain pipeline`, but newer did.

Maybe we can do it in both places, so that ids from explain will likely be the same

azat replied:
Will do this separately to avoid bloating this patch set.

KochetovNicolai commented:
Sorry for not reviewing it so long :(

azat commented:
Rebased, to fix conflicts.

azat commented:
>[Integration tests (asan, actions) [1/3] — fail: 1, passed: 669, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/34355/e5a837e574fac93e01d985f55d069fe522abc182/integration_tests__asan__actions__[1/3].html)

- `test_storage_postgresql_replica/test.py::test_abrupt_server_restart_while_heavy_replication` - flaky - #35885

azat committed:
Disable processors_profile_log for test_input_format_parallel_parsing_memory_tracking

This should fix memory usage, like in [1].

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/34355/e5a837e574fac93e01d985f55d069fe522abc182/integration_tests__thread__actions__[3/4].html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Stress test (thread, actions) — Fatal message in clickhouse-server.log (see fatal_messages.txt)](https://s3.amazonaws.com/clickhouse-test-reports/34355/acd48ebe2d51d20d0c2a5bc545d46d27f9bf7904/stress_test__thread__actions_.html)

The SIGSEGV that had been catched by gdb is for previous release, for `toString(toNullable(false)))` (that had been fixed in #35755), report will be fixed after #35985
